File: app/src/main/java/com/wa2c/android/medoly/plugin/action/tweet/dialog/InsertPropertyDialogFragment.java
package com.wa2c.android.medoly.plugin.action.tweet.dialog;

import android.app.Activity;
import android.app.AlertDialog;
import android.app.Dialog;
import android.content.Context;
import android.os.Bundle;
import android.view.View;
import android.view.ViewGroup;
import android.view.ViewGroup.LayoutParams;
import android.widget.AdapterView;
import android.widget.ArrayAdapter;
import android.widget.LinearLayout;
import android.widget.ListView;
import android.widget.TextView;

import com.wa2c.android.medoly.plugin.action.tweet.PropertyItem;
import com.wa2c.android.medoly.plugin.action.tweet.R;

import java.util.ArrayList;
import java.util.EventListener;


public class InsertPropertyDialogFragment extends AbstractDialogFragment {

    /**
     * ダイアログのインスタンスを作成する。
     * @return ダイアログのインスタンス。
     */
    static public InsertPropertyDialogFragment newInstance() {
        InsertPropertyDialogFragment fragment = new InsertPropertyDialogFragment();

        Bundle args = new Bundle();
        fragment.setArguments(args);
        return fragment;
    }



    /**
     * onCreateDialogイベント処理。
     */
    @Override
    public Dialog onCreateDialog(Bundle savedInstanceState) {
        final Activity context = getActivity();

        // 再生キュー
        itemList.addAll(PropertyItem.getDefaultPropertyPriority(context));

        // リストアダプタ
        final PropertytListAdapter adapter = new PropertytListAdapter(context, itemList);
        adapter.setNotifyOnChange(false);

        // リスト作成
        final ListView listView = new ListView(context);
        listView.setLayoutParams(new LayoutParams(LayoutParams.MATCH_PARENT, LayoutParams.WRAP_CONTENT));
        listView.setAdapter(adapter);
        listView.setFastScrollEnabled(true);
        listView.setOnItemClickListener(new AdapterView.OnItemClickListener() {
            @Override
            public void onItemClick(AdapterView<?> parent, View view, int position, long id) {
                if (itemSelectListener != null) {
                    String insert = "%" +itemList.get(position).propertyKey + "%";
                    itemSelectListener.onItemSelect(insert);
                }
                getDialog().dismiss();
            }
        });

        final LinearLayout listLayout = new LinearLayout(context);
        listLayout.setLayoutParams(new LayoutParams(LayoutParams.WRAP_CONTENT, LayoutParams.WRAP_CONTENT));
        listLayout.addView(listView);

        // ダイアログ
        AlertDialog.Builder builder = new AlertDialog.Builder(getActivity());
        builder.setTitle(R.string.label_dialog_insert_property_title);
        builder.setView(listLayout);

        // キャンセルボタン
        builder.setNegativeButton(android.R.string.cancel, null);

        return  builder.create();
    }

    /**
     * 項目リスト。
     */
    private final ArrayList<PropertyItem> itemList = new ArrayList<>();

    /**
     * リストアダプタ。
     */
    private class PropertytListAdapter extends ArrayAdapter<PropertyItem> {
        /** コンストラクタ。 */
        public PropertytListAdapter(Context context, ArrayList<PropertyItem> itemList) {
            super(context, android.R.layout.simple_list_item_1, itemList);
        }

        @Override
        public View getView(final int position, View convertView, ViewGroup parent) {
            final PropertyItem item = getItem(position);
            final ListItemViewHolder holder;

            if (convertView == null) {
                convertView = View.inflate(getActivity(), android.R.layout.simple_list_item_1, null);
                holder = new ListItemViewHolder();
                holder.TitleTextView = (TextView)convertView.findViewById(android.R.id.text1);
                convertView.setTag(holder);
            } else {
                holder = (ListItemViewHolder) convertView.getTag();
            }

            holder.TitleTextView.setText(item.propertyName);

            return convertView;
        }

        /** リスト項目のビュー情報を保持するHolder。 */
        class ListItemViewHolder {
            public TextView TitleTextView;
        }

    }




    /** 選択リスナ。 */
    private ItemSelectListener itemSelectListener;

    /**
     * 選択リスナを設定する。
     * @param listener リスナ。
     */
    public void setOnItemSelectListener(ItemSelectListener listener) {
        this.itemSelectListener = listener;
    }

    /**
     * 選択インターフェース。。
     */
    public interface ItemSelectListener extends EventListener {
        /**
         * 挿入文字を取得する。
         * @param insertString 挿入文字。
         */
        void onItemSelect(String insertString);
    }

}

File: app/src/main/java/com/wa2c/android/medoly/plugin/action/tweet/dialog/InsertPropertyDialogFragment.kt
package com.wa2c.android.medoly.plugin.action.tweet.dialog

import android.app.Dialog
import android.content.Context
import android.os.Bundle
import android.view.View
import android.view.ViewGroup
import android.view.ViewGroup.LayoutParams
import android.widget.*
import androidx.appcompat.app.AlertDialog
import com.wa2c.android.medoly.plugin.action.tweet.R
import com.wa2c.android.medoly.plugin.action.tweet.activity.PropertyItem
import java.util.*

/**
 * Insert property dialog.
 */
class InsertPropertyDialogFragment : AbstractDialogFragment() {

    /**  Item list. */
    private val itemList = ArrayList<PropertyItem>()

    /** Item select listener.  */
    var itemSelectListener: ItemSelectListener? = null

    /**
     * onCreateDialog
     */
    override fun onCreateDialog(savedInstanceState: Bundle?): Dialog {
        super.onCreateDialog(savedInstanceState)
        // List
        itemList.addAll(PropertyItem.getDefaultPropertyPriority(context))
        val listAdapter = PropertyListAdapter(context, itemList).apply {
            setNotifyOnChange(false)
        }

        // Create list view
        val listView = ListView(context).apply {
            layoutParams = LayoutParams(LayoutParams.MATCH_PARENT, LayoutParams.WRAP_CONTENT)
            adapter = listAdapter
            isFastScrollEnabled = true
            onItemClickListener = AdapterView.OnItemClickListener { _, _, position, _ ->
                val insert = "%" + itemList[position].propertyKey + "%"
                itemSelectListener?.onItemSelect(insert)
                dialog.dismiss()
            }
        }
        val listLayout = LinearLayout(context).apply {
            layoutParams = LayoutParams(LayoutParams.WRAP_CONTENT, LayoutParams.WRAP_CONTENT)
            addView(listView)
        }

        // Build dialog
        return AlertDialog.Builder(context).apply {
            setTitle(R.string.label_dialog_insert_property_title)
            setView(listLayout)
            setNeutralButton(android.R.string.cancel, null)
        }.create()
    }

    /**
     * List adapter
     */
    private inner class PropertyListAdapter(context: Context, itemList: ArrayList<PropertyItem>) : ArrayAdapter<PropertyItem>(context, android.R.layout.simple_list_item_1, itemList) {

        override fun getView(position: Int, convertView: View?, parent: ViewGroup): View {
            var view = convertView
            val holder: ListItemViewHolder

            if (view == null) {
                view = View.inflate(context, android.R.layout.simple_list_item_1, null)
                holder = ListItemViewHolder()
                holder.titleTextView = view!!.findViewById(android.R.id.text1) as TextView
                view.tag = holder
            } else {
                holder = view.tag as ListItemViewHolder
            }

            val item = getItem(position)
            if (item != null) {
                holder.titleTextView!!.text = item.propertyName
            }

            return view
        }

        /** リスト項目のビュー情報を保持するHolder。  */
        private inner class ListItemViewHolder {
            var titleTextView: TextView? = null
        }

    }



    /**
     * Item select listener
     */
    interface ItemSelectListener : EventListener {
        /**
         * Get insertion text.
         * @param insertString A insertion text.
         */
        fun onItemSelect(insertString: String)
    }

    companion object {
        /**
         * Create dialog instance.
         * @return New dialog instance.
         */
        fun newInstance(): InsertPropertyDialogFragment {
            return InsertPropertyDialogFragment().apply {
                arguments = Bundle()
            }
        }
    }

}
